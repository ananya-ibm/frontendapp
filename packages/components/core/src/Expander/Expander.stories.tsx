/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React from 'react';
import { Expander } from './Expander';

export default {
  title: 'Components/Core/Expander',
  component: Expander
};

const storyProps1 = {
  label: 'This is  a test'
};

const storyProps2 = {
  label: 'This is  a test',
  isDefaultExpanded: true
};

export const normal = args => (
  <Expander {...args}>
    <div>This is the expander content</div>
  </Expander>
);
normal.args = storyProps1;

export const Expanded = args => (
  <Expander {...args}>
    <div>This is the expander content</div>
  </Expander>
);
Expanded.args = storyProps2;
