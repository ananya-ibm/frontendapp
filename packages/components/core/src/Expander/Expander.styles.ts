/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import styled from 'styled-components';

export const Expander = styled('div')<{ isExpanded?: boolean }>`
  border-top: solid 0.0625rem ${props => props.theme.colors.delimiters.primary};
  display: flex;
  flex-direction: column;
  overflow-x: hidden;
  overflow-y: ${props => (props.isExpanded ? 'auto' : 'hidden')};
`;

export const Heading = styled('div')`
  cursor: pointer;
  display: flex;
  flex-grow: 1;
  font: 700 1rem/1.45rem ${props => props.theme.typography.body.short.M.family};
  padding-top: 0.5rem;
  position: relative;
`;

export const Content = styled('ol')<{ isExpanded?: boolean }>`
  list-style: none;
  max-height: ${props => (props.isExpanded ? '100%' : 0)};
  overflow-x: hidden;
  overflow-y: ${props => (props.isExpanded ? 'auto' : 'hidden')};
  transition: max-height 200ms ease-out;
`;

export const Icon = styled('div')`
  position: absolute;
  right: 0;
`;
