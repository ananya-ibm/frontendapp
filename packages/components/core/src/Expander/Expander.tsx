/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React, { useState } from 'react';
import { ChevronDown, ChevronUp } from '@carbon/react/icons';

import * as S from './Expander.styles';

export const Expander = ({ label, isDefaultExpanded, children }: Props) => {
  const [isExpanded, setIsExpanded] = useState(isDefaultExpanded);
  const onClick = () => {
    setIsExpanded(!isExpanded);
  };
  return (
    <>
      <S.Expander isExpanded={isExpanded}>
        <S.Heading data-testid="OpenMenu" onClick={onClick}>
          {label}
          <S.Icon> {isExpanded ? <ChevronUp size={24} /> : <ChevronDown size={24} />}</S.Icon>
        </S.Heading>
        <S.Content isExpanded={isExpanded}>{children}</S.Content>
      </S.Expander>
    </>
  );
};

type Props = {
  isDefaultExpanded?: boolean;
  label?: React.ReactElement | string;
  children?: any;
};
