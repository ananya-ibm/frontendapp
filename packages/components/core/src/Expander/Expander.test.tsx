/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React from 'react';
import { renderWithTheme as render } from '@testUtils';
import { fireEvent } from '@testing-library/react';
import { axe } from 'jest-axe';
import { Expander } from './Expander';

const testProps1 = {
  label: 'This is  a test',
  children: <li>Test content</li>
};

describe('<Expander /> component', () => {
  it('expands on heading click', () => {
    const { getByTestId, getByText } = render(<Expander {...testProps1} />);
    fireEvent.click(getByTestId('OpenMenu'));
    expect(getByText('Test content')).toBeDefined();
  });
  test('its snapshot matches', () => {
    const { container } = render(<Expander {...testProps1} />);

    expect(container).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();
  });
  test('it passes an axe accesibility check', async () => {
    const { container } = render(<Expander {...testProps1} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
