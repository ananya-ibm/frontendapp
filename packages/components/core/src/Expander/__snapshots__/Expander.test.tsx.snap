// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Expander /> component its snapshot matches 1`] = `
.c0 {
  border-top: solid 0.0625rem #8d8d8d;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-x: hidden;
  overflow-y: hidden;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font: 700 1rem/1.45rem 'IBM Plex Sans',Arial,sans-serif;
  padding-top: 0.5rem;
  position: relative;
}

.c3 {
  list-style: none;
  max-height: 0;
  overflow-x: hidden;
  overflow-y: hidden;
  -webkit-transition: max-height 200ms ease-out;
  transition: max-height 200ms ease-out;
}

.c2 {
  position: absolute;
  right: 0;
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="OpenMenu"
  >
    This is  a test
    <div
      class="c2"
    >
       
      <svg
        aria-label="icon"
        fill="currentColor"
        focusable="false"
        height="20"
        preserveAspectRatio="xMidYMid meet"
        role="img"
        size="24"
        viewBox="0 0 32 32"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </div>
  <ol
    class="c3"
  >
    <li>
      Test content
    </li>
  </ol>
</div>
`;
